File: app/src/main/java/com/beautifulquran/ui/theme/ContextualTipPlacement.kt
package com.beautifulquran.ui.theme

import androidx.compose.runtime.Immutable
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin

/**
 * Places teaching copy on a ray extending from the feature spotlight.
 *
 * Zero degrees points right and positive angles turn clockwise with screen Y.
 * [bodyDistanceFraction] is a share of the ray from the spotlight to the first
 * screen edge, so every angle remains inside the current surface.
 */
@Immutable
data class ContextualTipPlacement(
    val bodyAngleDegrees: Float,
    val bodyDistanceFraction: Float = 0.78f,
    val actionDistanceFraction: Float = 0.94f,
) {
    fun bodyCenter(spotlight: Offset, surface: Size): Offset =
        pointOnRay(spotlight, surface, bodyDistanceFraction)

    fun actionCenter(spotlight: Offset, surface: Size): Offset =
        pointOnRay(spotlight, surface, actionDistanceFraction)

    private fun pointOnRay(spotlight: Offset, surface: Size, fraction: Float): Offset {
        val radians = bodyAngleDegrees * (PI.toFloat() / 180f)
        val direction = Offset(cos(radians), sin(radians))
        val xLimit = when {
            direction.x > EPSILON -> (surface.width - spotlight.x) / direction.x
            direction.x < -EPSILON -> -spotlight.x / direction.x
            else -> Float.POSITIVE_INFINITY
        }
        val yLimit = when {
            direction.y > EPSILON -> (surface.height - spotlight.y) / direction.y
            direction.y < -EPSILON -> -spotlight.y / direction.y
            else -> Float.POSITIVE_INFINITY
        }
        val distance = min(xLimit, yLimit).coerceAtLeast(0f) * fraction.coerceIn(0f, 1f)
        return Offset(
            x = (spotlight.x + direction.x * distance).coerceIn(0f, surface.width),
            y = (spotlight.y + direction.y * distance).coerceIn(0f, surface.height),
        )
    }

    private companion object {
        const val EPSILON = 1e-4f
    }
}

File: web/src/ui/theme/contextualTipPlacement.ts
/**
 * Places teaching copy on a ray from the feature spotlight.
 *
 * Zero degrees points right; positive angles turn clockwise with screen Y —
 * Android `ContextualTipPlacement` parity.
 */

export type Point = { x: number; y: number }
export type Size = { width: number; height: number }

export type ContextualTipPlacement = {
  bodyAngleDegrees: number
  bodyDistanceFraction?: number
  actionDistanceFraction?: number
}

const EPSILON = 1e-4

function pointOnRay(
  spotlight: Point,
  surface: Size,
  angleDegrees: number,
  fraction: number,
): Point {
  const radians = (angleDegrees * Math.PI) / 180
  const dx = Math.cos(radians)
  const dy = Math.sin(radians)
  const xLimit =
    dx > EPSILON
      ? (surface.width - spotlight.x) / dx
      : dx < -EPSILON
        ? -spotlight.x / dx
        : Number.POSITIVE_INFINITY
  const yLimit =
    dy > EPSILON
      ? (surface.height - spotlight.y) / dy
      : dy < -EPSILON
        ? -spotlight.y / dy
        : Number.POSITIVE_INFINITY
  const distance =
    Math.max(0, Math.min(xLimit, yLimit)) * Math.min(1, Math.max(0, fraction))
  return {
    x: Math.min(surface.width, Math.max(0, spotlight.x + dx * distance)),
    y: Math.min(surface.height, Math.max(0, spotlight.y + dy * distance)),
  }
}

export function tipBodyCenter(
  placement: ContextualTipPlacement,
  spotlight: Point,
  surface: Size,
): Point {
  return pointOnRay(
    spotlight,
    surface,
    placement.bodyAngleDegrees,
    placement.bodyDistanceFraction ?? 0.78,
  )
}

export function tipActionCenter(
  placement: ContextualTipPlacement,
  spotlight: Point,
  surface: Size,
): Point {
  return pointOnRay(
    spotlight,
    surface,
    placement.bodyAngleDegrees,
    placement.actionDistanceFraction ?? 0.94,
  )
}

export function coercePoint(point: Point, surface: Size): Point {
  return {
    x: Math.min(surface.width, Math.max(0, point.x)),
    y: Math.min(surface.height, Math.max(0, point.y)),
  }
}

export function normalize(point: Point): Point {
  const length = Math.hypot(point.x, point.y)
  return length > EPSILON
    ? { x: point.x / length, y: point.y / length }
    : { x: 1, y: 0 }
}

export function midpoint(a: Point, b: Point): Point {
  return { x: (a.x + b.x) / 2, y: (a.y + b.y) / 2 }
}

export function dot(a: Point, b: Point): number {
  return a.x * b.x + a.y * b.y
}
